### PR TITLE
new auth class for SumoClient

### PR DESCRIPTION
**Pull request by adnejacobsen**
Added a new authentication class that supports two ways of acquiring a token: `device_flow` and `interactive`. The flow used is controlled by the `interactive` parameter which is `False` by default.

It will not make any attempts at acquiring a token on initialization like the previous `Auth` class, only when `get_token` is called. Token cache is only written back to file on program exit and not after each attempt at acquiring a token.

Currently only used in `SumoClient` when `token` parameter is not provided.

### CODE REVIEW

**Comment by adnejacobsen** [comment]
Can be tested like this:

```python
from sumo.wrapper import SumoClient

# interactive
sumo = SumoClient("dev", interactive=True)

result = sumo.get("/search", query="class:case", size=1, select=["_sumo"])
print(result)


# device code flow
sumo = SumoClient("dev")

result = sumo.get("/search", query="class:case", size=1, select=["_sumo"])
print(result)
```

Delete .sumo token cache before testing

File: src/sumo/wrapper/_new_auth.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                flow = self.msal.initiate_device_flow([self.scope])
+
+                if "user_code" not in flow:
+                    raise ValueError(
+                        "Fail to create device flow. Err: %s" % json.dumps(flow, indent=4)
+                    )
+
+                print(flow["message"])
+                result = self.msal.acquire_token_by_device_flow(flow)
+
```
Comment (rwiker) [inline src/sumo/wrapper/_new_auth.py:60]: Should there be a check on `"error" in result` here?

File: src/sumo/wrapper/_new_auth.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        if os.path.isfile(self.token_path):
+            with open(self.token_path, "r") as file:
+                cache.deserialize(file.read())
+
+        return cache
+
+    
+    def __save_cache(self):
+        if self.cache.has_state_changed:
+            old_mask = os.umask(0o077)
```
Comment (rwiker) [inline src/sumo/wrapper/_new_auth.py:76]: Should be 0o007, I think.

File: src/sumo/wrapper/_sumo_client.py [hunk 3]
```diff
@@ -1,63 +1,42 @@
...
 
     def _retrieve_token(self):
-        if self.user_provided_access_token:
-            self.logger.debug("User provided token exists, returning token")
-            return self.user_provided_access_token
-        else:
-            if self.auth.is_token_expired():
-                self.logger.debug("Token is expired, regenerating")
-                self.access_token = self.auth.get_token()
+        if self.access_token:
```
Comment (rwiker) [inline src/sumo/wrapper/_sumo_client.py:32]: Check whether access_token is expired?

**Review by rwiker — APPROVED** [review]
Looks good!

File: src/sumo/wrapper/_sumo_client.py (hunk 3 above)
Comment (adnejacobsen) [reply src/sumo/wrapper/_sumo_client.py:32]: Atm there is no logic for the user provided access_token, SumoClient assumes that a valid token has been provided. 

If the provided token expires before the SumoClient object dies, should we raise an exception? Or let the user provide a refresh token in the constructor and renew the access_token when it expires?

**Comment by rwiker** [comment]
I think I’d prefer to have the option of passing in either an access token or a refresh token. If a token is passed, we should bypass the authentication flows, and raise an exception if the access token is (becomes) unvalid, and we’re unable to refresh.

File: src/sumo/wrapper/_new_auth.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+                flow = self.msal.initiate_device_flow([self.scope])
+
+                if "error" in flow:
+                    raise ValueError(
+                        "Failed to create device flow. Err: %s" % json.dumps(flow, indent=4)
+                    )
+
+                print(flow["message"])
+                result = self.msal.acquire_token_by_device_flow(flow)
+
```
Comment (rwiker) [inline src/sumo/wrapper/_new_auth.py:60]: Need a check on result here, as acquire_token_by_device_flow may fail: [https://msal-python.readthedocs.io/en/latest/#msal.PublicClientApplication.acquire_token_by_refresh_token](https://msal-python.readthedocs.io/en/latest/#msal.PublicClientApplication.acquire_token_by_refresh_token)

File: src/sumo/wrapper/_new_auth.py [hunk 5]
```diff
@@ -58,6 +58,11 @@ def get_token(self):
                 print(flow["message"])
                 result = self.msal.acquire_token_by_device_flow(flow)
 
+                if "error" in result:
+                    raise ValueError(
+                        "Failed to create device flow. Err: %s" % json.dumps(flow, indent=4)
```
Comment (rwiker) [inline src/sumo/wrapper/_new_auth.py:63]: Flisespikring, men... "Failed to acquire token by device flow"?

Comment (adnejacobsen) [reply src/sumo/wrapper/_new_auth.py:63]: Stemmer!

Comment (adnejacobsen) [reply src/sumo/wrapper/_new_auth.py:63]: En litt rask copy/paste der..

**Review by thezultimate — APPROVED** [review]

**Pull request merged**